shoop/campaigns/models/campaign_effects.py

```diff
@@ -0,0 +1,110 @@
+# This file is part of Shoop.
+#
+# Copyright (c) 2012-2016, Shoop Ltd. All rights reserved.
+#
+# This source code is licensed under the AGPLv3 license found in the
+# LICENSE file in the root directory of this source tree.
+from enum import Enum
+
+from django.db import models
+from django.utils.translation import ugettext_lazy as _
+from polymorphic.models import PolymorphicModel
+
+from shoop.admin.forms.fields import PercentageField
+from shoop.core.fields import MoneyValueField
+from shoop.utils.properties import PriceProperty
+
+
+class EffectType(Enum):
+    CATALOG = 1
+    BASKET = 2
+
+
+class CampaignEffect(PolymorphicModel):
+    identifier = None
+    model = None
+    admin_form_class = None
+    yields = False
+
+    def set_locals(self, **kwargs):
+        self.order_source = kwargs.get("order_source")
+        self.price_info = kwargs.get("price_info")
+        self.context = kwargs.get("context")
+        self.product = kwargs.get("product")
+        self._campaign = kwargs.get("campaign")
+        self.set_type()
+
+    def set_type(self):
+        from shoop.campaigns.models import BasketCampaign
+        self.type = EffectType.BASKET if isinstance(self._campaign, BasketCampaign) else EffectType.CATALOG
+
+    def apply(self, **kwargs):
+        raise NotImplementedError("Not implemented!")
+
+    @property
+    def shop_currency(self):
+        return self._campaign.shop.currency
+
+    @property
+    def shop_prices_include_tax(self):
+        return self._campaign.shop.prices_include_tax
+
+
+class DiscountAmount(CampaignEffect):
+    identifier = "discount_amount_effect"
+    name = _("Discount amount value")
+
+    discount_amount = PriceProperty("discount_amount_value", "shop_currency", "shop_prices_include_tax")
```

Comment:
Since CampaignEffect is m2m to campaign, this just doesn't work. Or needs some hacks like the `set_locals` (which shouldn't be needed).

Wouldn't it be better to have the campaign FK in CampaignEffect? Then this property could simply be expressed as

```python
discount_amount = PriceProperty(
    "discount_amount_value",
    "campaign.shop.currency", "campaign.shop.prices_include_tax")
```

This would still allow having multiple effects for a campaign (if we want that later), but it would tie an effect to single campaign, which IMHO is a good thing.